txscript/engine.go

```diff
@@ -285,6 +285,36 @@ type Engine struct {
 	witnessProgram  []byte
 	inputAmount     int64
 	taprootCtx      *taprootExecutionCtx
+
+	// StepCallback is an optional function that will be called every time
+	// a step has been performed during script execution.
+	//
+	// NOTE: This is only meant to be used in debugging, and SHOULD NOT BE
+	// USED during regular operation.
+	StepCallback func(*StepInfo) error
+}
+
+// StepInfo houses the current VM state information that is passed back to the
+// StepCallback during script execution.
+type StepInfo struct {
+	// ScriptIndex is the index of the script currently being executed by
+	// the Engine.
+	ScriptIndex int
+
+	// OpcodeIndex is the index of the next opcode that will be executed.
+	OpcodeIndex int
+
+	// Witness is the witness of the current input.
+	//
+	// NOTE: this will be only be set before the witness program has been
+	// verified.
+	Witness [][]byte
```

Comment:
Is the witness actually needed? Given to create the VM in the first place, you had to have the witness. 

The other thing is that this field can actually change over time w/ the way execution works: ps2h, nested p2sh, segwit v0, taproot, etc.

Reply:
Yea, you're right. It's not really needed: https://github.com/btcsuite/btcd/pull/1980/commits/d1cb5debf4068853b282ad803e09602b63484301